### PR TITLE
YD-155 Added BudgetGoal subclass of Goal, fixed tests

### PR DESCRIPTION
**Pull request by aristotelos**

### CODE REVIEW

File: appservice/src/intTest/groovy/nu/yona/server/CreateUserOnBuddyRequestTest.groovy [hunk 1]
```diff
@@ -347,7 +347,9 @@ class CreateUserOnBuddyRequestTest extends AbstractAppServiceIntegrationTest
 				],
 				"goals":[
 					{
-						"activityCategoryName":"news"
+						"@class":"budgetGoal",
```
Comment (aristotelos) [inline appservice/src/intTest/groovy/nu/yona/server/CreateUserOnBuddyRequestTest.groovy]: Polymorphism has been implemented using a "@class" property. I am not sure how to do this better with HAL, I did not find good examples of it.

Comment (Bert-R) [inline appservice/src/intTest/groovy/nu/yona/server/CreateUserOnBuddyRequestTest.groovy]: Let's stick with this for now. We anyway need to handle goal changes differently, see [YD-165](http://jira.yona.nu/browse/YD-165).

File: core/src/main/java/nu/yona/server/goals/entities/BudgetGoal.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+@Entity
+public class BudgetGoal extends Goal
+{
+	private int maxDuration;
+
+	// Default constructor is required for JPA
+	public BudgetGoal()
+	{
+		super();
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/goals/entities/BudgetGoal.java]: This line isn't necessary.

File: core/src/main/java/nu/yona/server/goals/service/GoalDTO.java [hunk 3]
```diff
@@ -39,13 +37,15 @@ public String getActivityCategoryName()
...
 	}
 
-	public static GoalDTO createInstance(Goal entity)
-	{
-		return new GoalDTO(entity.getID(), entity.getActivityCategory().getName());
-	}
+	public abstract Goal createGoalEntity();
 
-	public Goal createGoalEntity()
+	public static GoalDTO createInstance(Goal g)
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/goals/service/GoalDTO.java]: Just use the full name: ``goal``.

**Comment by Bert-R** [comment]
I made two minor comments. For the rest it's fine.

File: appservice/src/intTest/groovy/nu/yona/server/CreateUserOnBuddyRequestTest.groovy (hunk 1 above)
Comment (aristotelos) [inline appservice/src/intTest/groovy/nu/yona/server/CreateUserOnBuddyRequestTest.groovy]: Agreed. It took me most time to implement the polymorphism like this but I realized it was lost time anyway :(. Maybe it is good to do YD-165 first or in combination with adding other goal types to take the impact as early as possible.

**Pull request merged**